content/docs/cucumber/state.md

```diff
@@ -238,6 +238,211 @@ compile group: 'io.cucumber', name: 'cucumber-needle', version: '{{% version "cu
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import com.google.inject.Stage;
+
+public final class CustomObjectFactory implements ObjectFactory {
+
+    private Injector injector;
+
+    public CustomObjectFactory() {
+        // Create an injector with our service module
+        this.injector =
+          Guice.createInjector( Stage.PRODUCTION, CucumberModules.createScenarioModule(), new ServiceModule());
+    }
+
+    @Override
+    public boolean addClass( Class< ? > clazz ) {
+        return true;
+    }
+
+    @Override
+    public void start() {
+        this.injector.getInstance( ScenarioScope.class ).enterScope();
+    }
+
+    @Override
+    public void stop() {
+        this.injector.getInstance( ScenarioScope.class ).exitScope();
+    }
+
+    @Override
+    public < T > T getInstance( Class< T > clazz ) {
+        return this.injector.getInstance( clazz );
+    }
+}
+```
+
+This is almost the default object factory for Guice except that we have added our own bindings to the injector.
+
+Cucumber loads the object factory through the `java.util.ServiceLoader`. In order for the ServiceLoader to be able
+to pick up our custom implementation we need to provide the file `META-INF/services/io.cucumber.core.backend.ObjectFactory`.
+
+```
+com.example.app.CustomObjectFactory
+#
+# ... additional custom object factories could be added here
+#
+```
+
+Our step definition should now be working except that we have to tell Cucumber to use our custom object factory.
+There are several ways how this could be accomplished.
+
+### The command line
+
+When Cucumber is run from the command line, the custom object factory can be specified as class name argument.
+
+```bash
+java io.cucumber.core.cli.Main --object-factory com.example.app.CustomObjectFactory
+```
+
+### The property file
+
+Cucumber makes use of a properties file (`cucumber.properties`) when it exists. The custom object factory can be
+specified in this file and will be picked up when Cucumber is running. The following entry needs to be available
+in the `cucumber.properties` file:
+
+```
+cucumber.object-factory=com.example.app.CustomObjectFactory
+```
+
+### The test runner (JUnit/TestNG)
+
+The Cucumber modules for JUnit and TestNG allow to run Cucumber through a JUnit/TestNG test. When the test is
+executed it actually starts the Cucumber engine that runs the Cucumber tests. Cucumber can be configured using
+the `@CucumberOptions` annotation. The custom object factory can be specified in this annotation.
+
+```java
+package com.example.app;
+
+import io.cucumber.junit.Cucumber;
+import io.cucumber.junit.CucumberOptions;
+import org.junit.runner.RunWith;
+
+@RunWith( Cucumber.class )
+@CucumberOptions( objectFactory = com.example.app.CustomObjectFactory.class )
+public final class RunCucumberTest {
+}
+```
+
+Using the `@CucumberOptions` annotation has another important advantage. It is possible to create different tests that
```

Comment:
Rather than duplicate information about @CucumberOptions here, lets link to https://cucumber.io/docs/cucumber/api/#junit for more information, eg:
"For more information about `@CucumberOptions` see [here]"(https://cucumber.io/docs/cucumber/api/#junit)

Reply:
I agree. This should actually be the place. As I have been going over my changes together with a collegue I came to the conclusion that I have tried to put a lot more of what I've learned about Cucumber in this doc addition. Some of it would be better places elsewhere.

Reply:
Ok, I have added more information to the JUnit description of the `@CucumberOption` annotation and linked it.